src/codegen/codegen_neuron_cpp_visitor.cpp

```diff
@@ -0,0 +1,723 @@
+        #include <math.h>
+        #include <stdio.h>
+        #include <stdlib.h>
+    )CODE");
+    if (!info.vectorize) {
+        printer->add_line("#include <vector>");
+    }
+}
+
+
+void CodegenNeuronCppVisitor::print_neuron_includes() {
+    printer->add_newline();
+    printer->add_multi_line(R"CODE(
+        #include "mech_api.h"
+        #include "neuron/cache/mechanism_range.hpp"
+        #include "nrniv_mf.h"
+        #include "section_fwd.hpp"
+    )CODE");
+}
+
+
+void CodegenNeuronCppVisitor::print_sdlists_init(bool print_initializers) {
+    for (auto i = 0; i < info.prime_variables_by_order.size(); ++i) {
+        const auto& prime_var = info.prime_variables_by_order[i];
+        /// TODO: Something similar needs to happen for slist/dlist2 but I don't know their usage at
+        // the moment
+        /// TODO: We have to do checks and add errors similar to nocmodl in the
+        // SemanticAnalysisVisitor
+        if (prime_var->is_array()) {
+            /// TODO: Needs a for loop here. Look at
+            // https://github.com/neuronsimulator/nrn/blob/df001a436bcb4e23d698afe66c2a513819a6bfe8/src/nmodl/deriv.cpp#L524
+            /// TODO: Also needs a test
+            printer->fmt_push_block("for (int _i = 0; _i < {}; ++_i)", prime_var->get_length());
+            printer->fmt_line("/* {}[{}] */", prime_var->get_name(), prime_var->get_length());
+            printer->fmt_line("_slist1[{}+_i] = {{{}, _i}}",
+                              i,
+                              position_of_float_var(prime_var->get_name()));
+            const auto prime_var_deriv_name = "D" + prime_var->get_name();
+            printer->fmt_line("/* {}[{}] */", prime_var_deriv_name, prime_var->get_length());
+            printer->fmt_line("_dlist1[{}+_i] = {{{}, _i}}",
+                              i,
+                              position_of_float_var(prime_var_deriv_name));
+            printer->pop_block();
+        } else {
+            printer->fmt_line("/* {} */", prime_var->get_name());
+            printer->fmt_line("_slist1[{}] = {{{}, 0}}",
+                              i,
+                              position_of_float_var(prime_var->get_name()));
+            const auto prime_var_deriv_name = "D" + prime_var->get_name();
+            printer->fmt_line("/* {} */", prime_var_deriv_name);
+            printer->fmt_line("_dlist1[{}] = {{{}, 0}}",
+                              i,
+                              position_of_float_var(prime_var_deriv_name));
+        }
+    }
+}
+
+
+void CodegenNeuronCppVisitor::print_mechanism_global_var_structure(bool print_initializers) {
+    /// TODO: Print only global variables printed in NEURON
+    printer->add_line();
+    printer->add_line("/* NEURON global variables */");
+    if (info.primes_size != 0) {
+        printer->fmt_line("static neuron::container::field_index _slist1[{0}], _dlist1[{0}];",
+                          info.primes_size);
+    }
+}
+
+
+/// TODO: Same as CoreNEURON?
```

Comment:
should be close, good idea to cross-check.

Reply:
Yes, I need to verify this again because I didn't remember if I validated this based on the logic of `nocmodl`

Reply:
It's indeed the same logic in `nocmodl` so this function can be moved to `CodegenCppVisitor`